File: wk1/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool checksum(long card);
void card_type(long card);
int first_two(long card);
int len(long card);

int main(void)
{
    long card = get_long("Number: ");
    bool check = checksum(card);
    if (check == 1)
    {
        card_type(card);
    }

    else
    {
        printf("INVALID\n");
    }

}

//Checksum returns sum
bool checksum(long card)
{
    int sum = 0;
    bool alternate = false;

    while (card >= 1)
    {
        int last_no = card % 10;
        if (alternate == true)
        {
            if (last_no >= 5)
            {
                sum += ((last_no * 2) % 10) + 1;
            }
            else
            {
                sum += last_no * 2;
            }

        }
        else if (alternate == false)
        {
            sum += last_no;
        }
        card /= 10;
        alternate = !alternate;
    }

    if (sum % 10 == 0)
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

//Card Type
void card_type(long card)
{
    //AMEX
    if (len(card) == 15 && (first_two(card) == 34 || first_two(card) == 37))
    {
        printf("AMEX\n");
    }

    //MC
    else if (len(card) == 16 && first_two(card) >= 51 && first_two(card) <= 55)
    {
        printf("MASTERCARD\n");
    }

    //VISA
    else if ((len(card) == 13 || len(card) == 16) && first_two(card) / 10 == 4)
    {
        printf("VISA\n");
    }
    //Invalid
    else
    {
        printf("INVALID\n");
    }
}

//First two digits
int first_two(long card)
{
    while (card >= 100)
    {
        card /= 10 ;
    }
    return card;
}


//Lenght of card
int len(long card)
{
    int length = 0;
    while (card >= 1)
    {
        card /= 10;
        length ++;

    }
    return length;
}

File: wk6/pset6/sentimental-credit/credit.py
# TODO

def main():
    card = int(input("Number: "))
    twos = str(card)[0] + str(card)[1]

    if checksum(card):

        if (len(str(card)) == 15 and (int(twos) == 34 or int(twos) == 37)):
            print("AMEX")

        #MC
        elif (len(str(card)) == 16 and int(twos) >= 51 and int(twos) <= 55):
            print("MASTERCARD")

        #VISA
        elif ((len(str(card)) == 13 or len(str(card)) == 16) and int(twos[0]) == 4):
            print("VISA")

        #Invalid
        else:
            print("INVALID")
    else:
        print("INVALID")

def checksum(card):
    sum = int(0)
    alternate = False
    while card > 1:
        last_no = int(card % 10)
        if alternate:
            if last_no >= 5:
                sum += ((last_no * 2) % 10 + 1)
            else:
                sum += last_no * 2
        else:
            sum += last_no
        card /= 10

        if alternate:
            alternate = False
        else:
            alternate = True


    if (sum % 10) == 0:
        return True
    else:
        return False

def card_type(card):
    twos = str(card)[0] + str(card)[1]

    if (len(str(card)) == 15 and (int(twos) == 34 or int(twos) == 37)):
        print("AMEX")

    #MC
    elif (len(str(card)) == 16 and int(twos) >= 51 and int(twos) <= 55):
        print("MASTERCARD")

    #VISA
    elif ((len(str(card)) == 13 or len(str(card)) == 16) and int(twos[0]) == 4):
        print("VISA")

    #Invalid
    else:
        print("INVALID")

    return None

if __name__ == "__main__":
    main()